File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
private:
    vector<string> ans;
    int n;
    vector<string> letters;
    string digits;
    
public:
    void backTracking(int i, string path) {
        if (path.size() == n) {
            ans.push_back(path);
            return;
        }
        for (char c : letters[digits[i] - '0' - 2]) {
            backTracking(i + 1, path + c);
        }
    }
    vector<string> letterCombinations(string digits) {
        this-> n = digits.size();
        if (n == 0) {
            return {};
        }
        int step = 3;
        char chr = 'a';
        for (int i = 2; i <= 9; ++i) {
            if (i == 7 or i == 9) {
                step = 4;
            }
            else {
                step = 3;
            }
            string letter;
            for (int j = 0; j < step; ++j) {
                letter += chr;
                chr += 1;
            }
            letters.push_back(letter);
        }
        this->digits = digits;
        backTracking(0, "");
        return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []
        
        n = len(digits)
        d = dict()
        c = 'a'
        step = 3
        
        for i in range(2, 10):
            string = ""
            if i == 7 or i == 9:
                step = 4
            else:
                step = 3
            for _ in range(step):
                string += c
                c = chr(ord(c) + 1)
            d[str(i)] = string
        
        self.ans = []
        
        def dfs(i, path):
            if i == n:
                self.ans.append(path)
                return
            for c in d[digits[i]]:
                dfs(i + 1, path + c)
               
        dfs(0, "")
        return self.ans